src/camera/camera_base_impl.h

```diff
@@ -0,0 +1,434 @@
+    }else{
+      return Undistort(ImagePlaneToWorld(clamped_pixel.cast<float>()));
+    }
+  }
+
+  template <typename Derived>
+  inline Eigen::Vector2f UnprojectFromImageCoordinates(const Eigen::MatrixBase<Derived>& pixel_position) const {
+    // Manual implementation of bilinearly filtering the lookup.
+    if(undistortion_lookup_){
+      Eigen::Vector2f clamped_pixel = pixel_position.cwiseMin(Eigen::Vector2f(width_ - 1.001f, height_ - 1.00f)).cwiseMax(0);
+      Eigen::Vector2i int_pos = clamped_pixel.cast<int>();
+      Eigen::Vector2f factor = clamped_pixel - int_pos.cast<float>();
+      Eigen::Vector2f top_left = undistortionLookup(int_pos.x(), int_pos.y());
+      Eigen::Vector2f top_right = undistortionLookup(int_pos.x() + 1, int_pos.y());
+      Eigen::Vector2f bottom_left = undistortionLookup(int_pos.x(), int_pos.y() + 1);
+      Eigen::Vector2f bottom_right = undistortionLookup(int_pos.x() + 1, int_pos.y() + 1);
+      return Eigen::Vector2f(
+          (1 - factor.y()) *
+                  ((1 - factor.x()) * top_left.x() + factor.x() * top_right.x()) +
+              factor.y() *
+                  ((1 - factor.x()) * bottom_left.x() + factor.x() * bottom_right.x()),
+          (1 - factor.y()) *
+                  ((1 - factor.x()) * top_left.y() + factor.x() * top_right.y()) +
+              factor.y() *
+                  ((1 - factor.x()) * bottom_left.y() + factor.x() * bottom_right.y()));
+    }else{
+      return Undistort(ImagePlaneToWorld(pixel_position));
+    }
+  }
+
+    // This iterative Undistort() function should not be used in
+  // time critical code. An undistortion texture may be preferable,
+  // as used by the UnprojectFromImageCoordinates() methods. Undistort() is only
+  // used for calculating this undistortion texture once.
+  template <typename Derived1, typename Derived2>
+  inline Eigen::Vector2f IterativeUndistort(const Eigen::MatrixBase<Derived1>& distorted_point,
+                                            const Eigen::MatrixBase<Derived2>& starting_point, bool* converged) const {
+    const std::size_t kNumUndistortionIterations = 100;
+    const Child* child = static_cast<const Child*>(this);
+
+    // Gauss-Newton.
+    const float kUndistortionEpsilon = 1e-10f;
+    if (converged) {
+      *converged = false;
+    }
+    Eigen::Vector2f undistorted = starting_point;
+    for (std::size_t i = 0; i < kNumUndistortionIterations; ++i) {
+      Eigen::Vector2f distorted_candidate = child->Distort(undistorted);
+      // (Non-squared) residuals.
+      Eigen::Vector2f delta_point = distorted_candidate - distorted_point;
+
+      // Accumulate H and b.
+      const Eigen::Matrix2f Jd = child->DistortionDerivative(undistorted);
+      const Eigen::Matrix2f Jd2 = Jd.transpose() * Jd;
+      const Eigen::Vector2f step = (Jd2.inverse() * Jd) * delta_point;
+      undistorted.noalias() -= step;
+      if (step.squaredNorm() < kUndistortionEpsilon) {
```

Comment:
Maybe a criteria on both would be good.
I can see that this was done for Benchmark camera, but e.g. polynomial camera did stop if the residual got worse instead (https://github.com/ETH3D/dataset-pipeline/blob/master/src/camera/camera_polynomial.h#L150) which caused problems.

The main reason I used step size instead of actual residual is because sometimes The point is not reachable, for example with high poynomail disortion cameras. The point here is not to considered the point to be found, but the convergence to be reached, be it at a wrong point or not, because once step size goes below a threshold, there is very little chance it will change with suplementary steps.

The ideal outcome of this algorithm would probably be three different states : not converged and not found (eg with fisheye), converged and not found, and converged and found. But this is a more general way, which suits e.g. colmap (https://github.com/colmap/colmap/blob/dev/src/base/camera_models.h#L547), while here, we make the assumptions that the camera models is reasonnable, and having a polynomail model camera that has a radius cutoff below image radius is clearly wrong and should be replaced with a fisheye model. Not gonna lie, I balantly followed COLMAP's own undistortion code, but i see now how it doesn't apply as well here.

Reply:
Okay, I see that this change helps avoiding to go to the maximum iteration count if the point is actually unreachable. I guess it's fine if it works with the kinds of reasonable models that are used in practice.

Perhaps it would still be better though to base the decision of whether to set `*converged` to true on whether the cost actually reached near-zero (with a relaxed threshold)? I think this might usually be the more useful piece of information for the caller of this function, since one probably usually wants to know whether the return value is a valid result (this is if the cost is near-zero), or if it is an invalid result (that is if the optimization stopped early, or it stopped but couldn't reach the goal)?